File: java/leetcode/src/main/java/com/leetcode/graph/CourseSchedule_207.java
package com.leetcode.graph;

import java.util.*;

public class CourseSchedule_207 {
    private boolean canfinish = true;

    public boolean canFinish(int numCourses, int[][] prerequisites) {

        dfs_solution(numCourses, prerequisites);
        //canfinish = bfs_solution(numCourses, prerequisites);

        return canfinish;
    }

    private boolean bfs_solution(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int[] pre : prerequisites) {
            if (!map.containsKey(pre[1]))
                map.put(pre[1], new ArrayList<Integer>());
            map.get(pre[1]).add(pre[0]);
            indegree[pre[0]]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        int removed = 0;

        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.offer(i);
                removed++;
            }
        }

        while (!queue.isEmpty()) {
            Integer cur = queue.poll();
            if (map.containsKey(cur)) {
                for(int c : map.get(cur)) {
                    indegree[c]--;
                    if (indegree[c] == 0) {
                        queue.offer(c);
                        removed++;
                    }
                }
            }

        }

        return numCourses == removed;
    }


    private void dfs_solution(int numCourses, int[][] prerequisites) {
        boolean[] visited = new boolean[numCourses];
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int[] pre: prerequisites) {
            if (!map.containsKey(pre[1])) {
                map.put(pre[1], new ArrayList<Integer>());
            }
            map.get(pre[1]).add(pre[0]);
        }

        for (int i = 0; i < numCourses; i++) {
            if (!visited[i])
                dfs(i, map, visited, new ArrayList<Integer>());
            if (!canfinish) return;
        }

    }

    private void dfs(Integer course, Map<Integer,List<Integer>> map, boolean[] visited, List<Integer> path) {
        visited[course] = true;
        path.add(course);
        if (map.containsKey(course)) {
            for (Integer c : map.get(course)) {
                if (!canfinish) return;
                if (!visited[c]) {
                    dfs(c, map, visited, path);
                } else {
                    if (path.contains(c)) {
                        canfinish = false;
                        return;
                    }
                }
            }
        }
        path.remove(course);
    }

    public static void main(String[] args) {
        int[][] preqs = {{0, 1}};
        CourseSchedule_207 courseSchedule_207 = new CourseSchedule_207();
        boolean res  = courseSchedule_207.canFinish(2, preqs);
        System.out.println(res);

    }
}

File: js/algrithm/Graph/Course_Schedule_207.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    //const res = bfs(numCourses, prerequisites)
    const res = dfs(numCourses, prerequisites)
    return res
};

function dfs(numCourses, prerequisites) {
    const digraph = []
    for (let pre of prerequisites) {
        if (!digraph[pre[1]]) digraph[pre[1]] = []
        digraph[pre[1]].push(pre[0])
    }

    const visited = Array(numCourses).fill(0)
    for (let i = 0; i < numCourses; i++) {
        if (visited[i] === 0) {
            if(!_dfs(digraph, i, visited))  return false
        }
        
    }
    return true
}
function _dfs(digraph, visited, i, memo) {
    if (visited[i] === 1) return false
    if (visited[i] === 2) return true
    visited[i] = 1
    if (digraph[i]) {
        for (let next of digraph[i]) {
            if (visited[next] === 1) return false
            if (visited[next] === 0) {
                if (!_dfs(digraph, next, visited)) return false
            }
        }
    }
    visited[i] = 2
    return true
}

function bfs(numCourses, prerequisites) {
    const digraph = []
    const degree = Array(numCourses).fill(0)
    const queue = []
    let cnt = 0

    for(let pre of prerequisites) {
        if (!digraph[pre[1]]) {
            digraph[pre[1]] = []
        }
        digraph[pre[1]].push(pre[0])
        degree[pre[0]]++
    }
    for (let i = 0; i < numCourses; i++) {
        if (degree[i] === 0) {
            queue.push(i)
        }
    }

    while (queue.length) {
        const cur = queue.shift()
        if (digraph[cur]) {
            for (let next of digraph[cur]) {
                if (--degree[next] === 0) {
                    queue.push(next)
                }
            }
        }
        cnt++
    }

    return cnt === numCourses
}
